src/Microsoft.ML.Data/DataView/CacheDataView.cs

```diff
@@ -682,7 +682,11 @@ public bool Wait(long pos)
             {
                 foreach (var w in _waiters)
                     w.Wait(pos);
-                return pos < _parent._rowCount || _parent._rowCount == -1;
+
+                // _parent._rowCount may or may not be initialized at this point. Only read the value once
+                // to avoid race conditions.
+                var rowCount = _parent._rowCount;
+                return rowCount == -1 || pos < rowCount;
             }
```

Comment:
Great job figuring this out!! 
As discussed over chat, please move this fix to inside GetRowCount() and use GetRowCount everywhere throughout the file wherever _rowCount is used outside the CacheDataViewClass.